File: test/interop/ts/src/legacy_sse_client.ts
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js';

class ProtocolTrackingSSEClientTransport extends SSEClientTransport {
  protocolVersion?: string;

  override setProtocolVersion(version: string): void {
    this.protocolVersion = version;
    super.setProtocolVersion(version);
  }
}

function readArg(args: string[], name: string): string | undefined {
  const index = args.indexOf(name);
  if (index < 0 || index + 1 >= args.length) {
    return undefined;
  }
  return args[index + 1];
}

async function main(): Promise<void> {
  const url = readArg(process.argv.slice(2), '--url');
  if (!url) {
    throw new Error('--url is required');
  }

  const client = new Client(
    { name: 'ts-legacy-sse-client', version: '1.0.0' },
    { capabilities: {} }
  );
  const transport = new ProtocolTrackingSSEClientTransport(new URL(url));

  try {
    await client.connect(transport);
    if (transport.protocolVersion !== '2025-11-25') {
      throw new Error(
        `Unexpected negotiated protocol: ${transport.protocolVersion}`
      );
    }
    const server = client.getServerVersion();
    if (server?.name !== 'example-dart-server') {
      throw new Error(
        `Unexpected Dart server identity: ${JSON.stringify(server)}`
      );
    }

    const tools = await client.listTools();
    if (!tools.tools.some((tool) => tool.name === 'calculate')) {
      throw new Error(
        `Dart server did not advertise calculate: ${JSON.stringify(tools)}`
      );
    }

    const result = await client.callTool({
      name: 'calculate',
      arguments: { operation: 'add', a: 20, b: 22 },
    });
    const content = result.content as Array<{ type: string; text?: string }>;
    if (content[0]?.type !== 'text' || content[0].text !== 'Result: 42') {
      throw new Error(`Unexpected calculate result: ${JSON.stringify(result)}`);
    }

    console.log('TypeScript legacy SSE client interop passed: Result: 42');
  } finally {
    await client.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: test/interop/python_2026_07_28/legacy_sse_client.py
import argparse
import asyncio
import json

from mcp import ClientSession
from mcp.client.sse import sse_client


async def run(url: str) -> None:
    async with sse_client(url) as (read, write):
        async with ClientSession(read, write) as session:
            initialized = await session.initialize()
            if initialized.protocol_version != "2025-11-25":
                raise RuntimeError(
                    "Expected protocol 2025-11-25, got "
                    f"{initialized.protocol_version}"
                )
            if initialized.server_info.name != "example-dart-server":
                raise RuntimeError(
                    f"Unexpected Dart server identity: {initialized.server_info!r}"
                )

            tools = await session.list_tools()
            names = {tool.name for tool in tools.tools}
            if "calculate" not in names:
                raise RuntimeError(f"Expected calculate tool, got {sorted(names)}")

            result = await session.call_tool(
                "calculate",
                {"operation": "add", "a": 20, "b": 22},
            )
            text = getattr(result.content[0], "text", None)
            if text != "Result: 42":
                raise RuntimeError(f"Unexpected calculate result: {result!r}")

            print(
                json.dumps(
                    {
                        "protocolVersion": initialized.protocol_version,
                        "serverName": initialized.server_info.name,
                        "toolCount": len(tools.tools),
                        "result": text,
                    }
                ),
                flush=True,
            )


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--url", required=True)
    args = parser.parse_args()
    asyncio.run(run(args.url))


if __name__ == "__main__":
    main()
